lib/template-loader.php

```diff
@@ -0,0 +1,162 @@
+<?php
+/**
+ * Block template loader functions.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Adds necessary filters to use 'wp_template' posts instead of theme template files.
+ */
+function gutenberg_add_template_loader_filters() {
+	if ( ! post_type_exists( 'wp_template' ) ) {
+		return;
+	}
+
+	/**
+	 * Array of all overrideable default template types.
+	 *
+	 * @see get_query_template
+	 *
+	 * @var array
+	 */
+	$template_types = array(
+		'index',
+		'404',
+		'archive',
+		'author',
+		'category',
+		'tag',
+		'taxonomy',
+		'date',
+		// Skip 'embed' for now because it is not a regular template type.
+		'home',
+		'frontpage',
+		'privacypolicy',
+		'page',
+		'search',
+		'single',
+		'singular',
+		'attachment',
+	);
+	foreach ( $template_types as $template_type ) {
+		add_filter( $template_type . '_template', 'gutenberg_override_query_template', 20, 3 );
+	}
+
+	add_filter( 'template_include', 'gutenberg_find_template', 20 );
+}
+add_action( 'wp_loaded', 'gutenberg_add_template_loader_filters' );
+
+/**
+ * Filters into the "{$type}_template" hooks to record the current template hierarchy.
+ *
+ * The method returns an empty result for every template so that a 'wp_template' post
+ * is used instead.
+ *
+ * @see gutenberg_find_template
+ *
+ * @param string $template  Path to the template. See locate_template().
+ * @param string $type      Sanitized filename without extension.
+ * @param array  $templates A list of template candidates, in descending order of priority.
+ * @return string Empty string to ensure template file is considered not found.
+ */
+function gutenberg_override_query_template( $template, $type, array $templates = array() ) {
+	global $_wp_current_template_hierarchy;
+
+	if ( ! is_array( $_wp_current_template_hierarchy ) ) {
+		$_wp_current_template_hierarchy = $templates;
+	} else {
+		$_wp_current_template_hierarchy = array_merge( $_wp_current_template_hierarchy, $templates );
```

Comment:
When will this ever run more than once and require this merging?

And what guarantees subsequent runs pass `$templates` where the first item is of lesser priority than the last item of the current `$_wp_current_template_hierarchy`?

Reply:
Based on how core's template loader works, this _may_ be run multiple times (e.g. `get_index_template()` being called after another `get_*_template()` function). As well based on core, the functions are called in the correct hierarchy order (more specific come first), so it will always be correct regarding priority.

Note that this is only a layer to make it work in the plugin anyway. Most likely, if/when this gets merged into core, it would be more directly integrated with the existing template loader code.

Reply:
Thanks, that makes sense.